base/clistruct/clistruct.go

```diff
@@ -0,0 +1,224 @@
+
+		// Handle embedded structs without nesting struct tag values
+		if valTypeField.Type.Kind() == reflect.Struct && valTypeField.Anonymous {
+			if err := register(fs, val.Field(i), val.Field(i).Type(), namePrefix, structNameTag); err != nil {
+				return err
+			}
+			continue
+		}
+
+		flagName := valTypeField.Tag.Get(structNameTag)
+
+		// Handle JSON tags as a special case for the options and struct field fallback handling.
+		if structNameTag == jsonStructNameTag {
+			var skip bool
+			flagName, skip = removeJSONTagOpts(flagName)
+			if skip {
+				continue
+			}
+			if flagName == "" {
+				// json.Marshal takes the struct field name in this case
+				flagName = valTypeField.Name
+			}
+		}
+
+		// skip untagged fields
+		if flagName == "" {
+			continue
+		}
+
+		// nest the given flagName under the parent tags
+		if namePrefix != "" {
+			flagName = namePrefix + "." + flagName
+		}
+
+		fieldType := valTypeField.Type
+		fieldValue := val.Field(i)
+
+		if fieldType.Kind() == reflect.Ptr && fieldType.Elem().Kind() == reflect.Struct {
+			// deref pointers to structs in order to recurse
+			fieldType = fieldType.Elem()
+			if fieldValue.IsNil() {
+				fieldValue.Set(reflect.New(fieldType))
+			}
+			fieldValue = fieldValue.Elem()
+		}
+
+		if fieldType.Kind() == reflect.Struct {
+			// recurse for nested structs
+			if err := register(fs, fieldValue, fieldType, flagName, structNameTag); err != nil {
+				return err
+			}
+			continue
+		}
+
+		if !fieldValue.CanSet() {
+			return fmt.Errorf("value %q (%s) can't be set - not addressable or unexported", valTypeField.Name, structNameTag)
+		}
+
+		flagHelp := valTypeField.Tag.Get(cliStructHelpTag)
+		if err := assignFlagToStructFieldVar(fs, fieldValue, flagName, flagHelp); err != nil {
+			return err
+		}
+
+	}
+	return nil
+}
+
+// assignFlagToStructFieldVar creates a flag associated with the given struct field, if the type can be set from a flag.
+func assignFlagToStructFieldVar(fs *flag.FlagSet, fieldValue reflect.Value, flagName, flagHelp string) error {
+
+	// get the pointer for non-pointer fields
+	if fieldValue.Kind() != reflect.Ptr && fieldValue.CanAddr() {
+		fieldValue = fieldValue.Addr()
+	}
+
+	// The assignments below take 2 forms:
+	// If v is nil, create a flag var and assign it to the struct
+	// If v is not nil, bind the flag to the variable in the struct
+
+	i := fieldValue.Interface()
+
+	// wrappers for flag.Value/fs.Var handling
+	switch v := i.(type) {
+	// Implement wrapper types in flag_value_wrappers.go for common types and wrap them here.
+	case *[]string:
+		if v == nil {
+			tmp := make(stringSliceValue, 0)
+			i = &tmp
+			fieldValue.Set(reflect.ValueOf((*[]string)(&tmp)))
+		} else {
+			i = (*stringSliceValue)(v)
+		}
+	case *json.Number:
+		if v == nil {
+			tmp := jsonNumberFlagValue("0")
+			i = &tmp
+			fieldValue.Set(reflect.ValueOf((*json.Number)(&tmp)))
+		} else {
+			i = (*jsonNumberFlagValue)(v)
+		}
+	}
+
+	switch v := i.(type) {
```

Comment:
I don't completely follow why we need a second type switch here, as opposed to just adding the fs.Var to the cases above.

Reply:
No reason over having multiple calls to `fs.Var` - changed as it's easier to reason about when explicit.